validate/validate.go

```diff
@@ -0,0 +1,574 @@
+package validate
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"unicode"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/blang/semver"
+	rspec "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+var (
+	defaultRlimits = []string{
+		"RLIMIT_CPU",
+		"RLIMIT_FSIZE",
+		"RLIMIT_DATA",
+		"RLIMIT_STACK",
+		"RLIMIT_CORE",
+		"RLIMIT_RSS",
+		"RLIMIT_NPROC",
+		"RLIMIT_NOFILE",
+		"RLIMIT_MEMLOCK",
+		"RLIMIT_AS",
+		"RLIMIT_LOCKS",
+		"RLIMIT_SIGPENDING",
+		"RLIMIT_MSGQUEUE",
+		"RLIMIT_NICE",
+		"RLIMIT_RTPRIO",
+		"RLIMIT_RTTIME",
+	}
+	defaultCaps = []string{
+		"CAP_CHOWN",
+		"CAP_DAC_OVERRIDE",
+		"CAP_FSETID",
+		"CAP_FOWNER",
+		"CAP_MKNOD",
+		"CAP_NET_RAW",
+		"CAP_SETGID",
+		"CAP_SETUID",
+		"CAP_SETFCAP",
+		"CAP_SETPCAP",
+		"CAP_NET_BIND_SERVICE",
+		"CAP_SYS_CHROOT",
+		"CAP_KILL",
+		"CAP_AUDIT_WRITE",
+	}
+)
+
+type Validator struct {
+	spec         rspec.Spec
+	rootfs       string
```

Comment:
On Thu, Sep 08, 2016 at 12:09:20AM -0700, 梁辰晔 (Liang Chenye) wrote:
> > +type Validator struct {
> > +	spec         rspec.Spec
> > +	rootfs       string
>
> I'm think it will be tricky if a user changes `BundlePath`, `Spec`
> need to be updated accordingly.  So I think `Generator` is right, it
> only publics `HostSpecific`.

I don't see why you'd need to update Spec if you change BundlePath.
Can you elaborate?  Maybe the user wants to validate several rootfses
(or whatever) against the same config, so they:

  validator := &Validator{
    Spec: spec,
    HostSpecific: true
  }
  for _, bundlePath := range bundles {
    validator.BundlePath = bundlePath
    msg = append(msgs, validator.CheckAll())
  }